multicast/discovery.go

```diff
@@ -57,13 +59,14 @@ func (d *Discovery) Respond(ctx context.Context, info ServerInfo) error {
 		return fmt.Errorf("Failed to listen on %d: %w", d.port, err)
 	}
 
-	receiverP := ipv4.NewPacketConn(receiver)
-	err = receiverP.JoinGroup(iface, &net.UDPAddr{IP: d.group})
+	ctx, d.responderCancel = context.WithCancel(ctx)
```

Comment:
Overwriting the ctx can lead to confusion and potential bugs, especially if the original context carries important deadlines, cancellations, or values that are relied upon elsewhere in the code (maybe there are none and then this is fine). Any cancellation or deadline on the original ctx will still propagate, but overwriting it here means that any modifications to ctx after this point will not affect the original context. Instead of overwriting ctx, we could introduce a new variable for the child context:
```go
responderCtx, responderCancel := context.WithCancel(ctx)
d.responderCancel = responderCancel
```

Then, use `responderCtx` within the goroutine down below. This approach maintains the integrity of the original ctx and also makes it easier to know which context is being used.

Reply:
Isn't this the go-to pattern for deriving new contexts if you don't care about the parent context? See also the docs of [`WithTimeout`](https://pkg.go.dev/context#WithTimeout).

In this case a new context is derived from the outer context and used internally so we can cancel it ourselves (as we don't control the outer context) any time. And if the parent context gets cancelled also our derived child context exits.

Reply:
Yes true. Nevermind then :)